src/ios/MultipeerConnectivity/THEMultipeerSocketRelay.m

```diff
@@ -0,0 +1,254 @@
+  assert(_socket == nil);
+
+  _socket = socket;
+  [self openStreams];
+  [_socket readDataWithTimeout:-1 tag:0];
+}
+
+- (void)stop
+{
+  NSLog(@"%@ relay: stopping", _relayType);
+
+  if (_socket)
+  {
+    NSLog(@"%@ relay: destroying socket %p", _relayType, _socket);
+    _socket.delegate = nil;
+    [_socket disconnect];
+    _socket = nil;
+  }
+
+  if (_inputStream)
+  {
+    [_inputStream close];
+    [_inputStream removeFromRunLoop: [NSRunLoop currentRunLoop] forMode: NSDefaultRunLoopMode];
+    _inputStream = nil;
+  }
+
+  if (_outputStream)
+  {
+    [_outputStream close];
+    [_outputStream removeFromRunLoop: [NSRunLoop currentRunLoop] forMode: NSDefaultRunLoopMode];
+    _outputStream = nil;
+  }
+}
+
+- (void)dealloc
+{
+  NSLog(@"%@ relay: dealloc", _relayType);
+  [self stop];  
+}
+
+- (void)socket:(GCDAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag
+{
+  assert(sock == _socket);
+  assert(_outputStream != nil);
+
+  if ([_outputStream write:data.bytes maxLength:data.length] != data.length)
+  {
+    NSLog(@"ERROR: Writing to output stream");
+  }
+
+  [_socket readDataWithTimeout:-1 tag:tag];
+}
+
+- (void)socket:(GCDAsyncSocket *)sock didWriteDataWithTag:(long)tag
+{
+}
+
+- (void)socketDidDisconnect:(GCDAsyncSocket *)sock withError:(NSError *)err
+{
+  assert(_socket == nil || sock == _socket);
+
+  // Usually benign, the upper layer just closed their connection
+  // they may want to connect again later
+
+  NSLog(@"%@ relay: socket disconnected", _relayType);
+
+  if (err) 
+  {
+      NSLog(@"%@ relay: %p disconnected with error %@ ", _relayType, sock, [err description]);
+  }
+}
+
+#pragma mark - NSStreamDelegate
+
+- (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode
+{
+  if (aStream == _inputStream) 
+  {
+    switch (eventCode) 
+    {
+      case NSStreamEventOpenCompleted:
+      {
+        //NSLog(@"%@ relay: inputStream opened", _relayType);
+      }
+      break;
+
+      case NSStreamEventHasSpaceAvailable:
+      {
+        //NSLog(@"%@ relay: inputStream hasSpace", _relayType);
+      }
+      break;
+
+      case NSStreamEventHasBytesAvailable:
+      {
+        const uint bufferValue = 1024;
+        uint8_t *buffer = malloc(bufferValue);
+        NSInteger len = [_inputStream read:buffer maxLength:sizeof(bufferValue)];
+        if (len)
+        {
+          NSMutableData *toWrite = [[NSMutableData alloc] init];
+          [toWrite appendBytes:buffer length:len];
+
+          assert(_socket);
+          [_socket writeData:toWrite withTimeout:-1 tag:len];
+        }
+      }
+      break;
+
+      case NSStreamEventEndEncountered:
+      {
+        //NSLog(@"%@ relay: inputStream closed", _relayType);
+      }
+      break;
+
+      case NSStreamEventErrorOccurred:
+      {
+        //NSLog(@"%@ relay: inputStream error", _relayType);
+      }
+      break;
+
+      default:
+      {
+      }
+      break;
+    }
+  }
+  else if (aStream == _outputStream)
```

Comment:
Is handy when you need to debug an issue to be reminded right there on the page what could happen and to have handy log messages to uncomment to quickly figure stuff out.

Doesn't offend me.